src/test/java/com/google/javascript/clutz/this_type_param.d.ts

```diff
@@ -2,7 +2,8 @@ declare namespace ಠ_ಠ.clutz.nsThis {
   class C {
     private noStructuralTyping_: any;
     bar ( ) : ಠ_ಠ.clutz.nsThis.C ;
-    foo ( ) : this ;
+    foo < THIS > (this : THIS ) : THIS ;
```

Comment:
I think it might be slightly preferable to emit an actual `this` type here.

In TypeScript, if you extend this class, it'll be rather cumbersome to create the overridden method - you'll need to deal with all the template args etc instead of just `foo(): this { ... }`. That syntax is also a bit more readable to TS users.

That said, if that's hard to implement or has other corner cases, I think it's OK to go with this - I expect few people will need to extend `<THIS>` templated base classes written in Closure from TypeScript.

Reply:
@mprobst I think so. I've tried to handle special simple case. https://github.com/angular/clutz/pull/802/commits/89f4f2b99cbd1cf021e9cd0949cac51b7eab5e47